lib/src/bitarray.dart

```diff
@@ -0,0 +1,317 @@
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    5, 6, 6, 7, 6, 7, 7, 8
+  ];
+
+  /// Creates a [BitArray] with [length].
+  ///
+  /// All initial values are [false], set [filled] to [true] to make them [true].
+  BitArray(int length, [bool filled = false]) {
+    ArgumentError.checkNotNull(length);
+    RangeError.checkNotNegative(length, 'length');
+
+    _length = length;
+    _data = Uint32List((length >> 5) + 1);
+
+    _tailMask = 0xFFFFFFFF >> (_data.length * 32 - _length);
+
+    if (filled) {
+      _setDataWords((_) => -1);
+    }
+  }
+
+  /// Generates a [BitArray] filled with values by [generator].
+  factory BitArray.generate(int length, bool Function(int index) generator) {
+    ArgumentError.checkNotNull(length);
+    RangeError.checkNotNegative(length, 'length');
+
+    ArgumentError.checkNotNull(generator);
+
+    var result = BitArray(length);
+    for (var pos = 0; pos < length; ++pos) {
+      result[pos] = generator(pos);
+    }
+    return result;
+  }
+
+  /// Replaces integer values in [_data] by applying [func] to each word.
+  void _transformDataWords(int Function(int word, int wordIndex) func) {
+    for (var wordIndex = 0; wordIndex < _data.length; ++wordIndex) {
+      _data[wordIndex] = func(_data[wordIndex], wordIndex);
+    }
+    _clearTailBits();
+  }
+
+  /// Sets all words in [_data] to values retuned from [func].
+  void _setDataWords(int Function(int wordIndex) func) {
+    for (var wordIndex = 0; wordIndex < _data.length; ++wordIndex) {
+      _data[wordIndex] = func(wordIndex);
+    }
+    _clearTailBits();
+  }
+
+  /// Sets tail bits in the last word of [_data] to zero.
+  ///
+  /// Needs to clear last bits after logical operations over [this].
+  void _clearTailBits() {
+    _data.last &= _tailMask;
+  }
+
+  /// Performs logical negation to [this].
+  void not() {
+    _transformDataWords((int word, _) => ~word);
+  }
+
+  /// Performs logical AND operation to [this] with given [BitArray].
+  void and(BitArray ba) {
+    _checkLengthsEqual(this, ba);
+    _transformDataWords(
+        (int word, int wordIndex) => word & ba._data[wordIndex]);
```

Comment:
I think these higher-order functions are going to be a performance problem (creating closures that might not be fully inlined).

I would write this body as a loop, and as there are several paths that have a loop doing '&', make that loop be a private static method called from all places
The call
```dart
    _transformDataWords(
        (int word, int wordIndex) => word & ba._data[wordIndex]);
```
would then be replace by
```dart
   _bitwiseAnd(this._data, this._data, ba._data);
```
```dart
// [target] can be the same list as [source1] or [source2].
static void _bitwiseAnd(Uint32List target, Uint32List source1, Uint32List source2) {
  assert(target.length <= source1.length && target.length <= source2.length);
  for (int i = 0; i < target.length; i++) {
    target[i] = source1[i] & source2[i];
  }
  // no tailmask required for '&', but could be passed in.
}
```

Reply:
These functions will be replaced with at place loops.